File: 04-distributed/01-partial-failure-and-the-ambiguous-result/golang/ambiguous_result.go
// Layer 4 · Topic 1 — the third outcome, in Go.
//
// WHAT THIS DEMONSTRATES
//
//	The same five faults as the Python and Node programs, against an in-process
//	ledger that records server-side truth, so the client's belief can be diffed
//	against what actually happened.
//
//	Go's contribution is net/http/httptrace. Every other client in this lab has
//	to infer which phase a failure happened in; Go can be told. ConnectDone and
//	WroteRequest fire at exactly the boundary that decides whether a retry is
//	safe, which makes the classification a fact rather than a guess.
//
// WHAT TO LOOK FOR
//  1. Phase 1's duplicate charges: created by the client's own retries, not by
//     the faults.
//  2. Phase 2's duplicates (gone) next to its unresolved ambiguity (unchanged).
//  3. The note at the end about context cancellation. Cancelling the client's
//     context tears down *your* side. The handler that already read the request
//     commits anyway, which is why the ledger keeps growing while the caller
//     believes everything failed.
//
// Run:  go run ambiguous_result.go
package main

import (
	"context"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/http/httptrace"
	"sort"
	"strings"
	"sync"
	"syscall"
	"time"
)

const (
	clientTimeout   = 300 * time.Millisecond // the deadline the caller will wait
	slowResponse    = 1000 * time.Millisecond
	requestsPerMode = 4
	maxAttempts     = 3
)

var modes = []string{"ok", "slow", "hang", "reset", "crash_after_commit", "refused"}

// --- server-side truth ------------------------------------------------------

type ledger struct {
	mu   sync.Mutex
	rows []string
}

func (l *ledger) commit(chargeID string) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.rows = append(l.rows, chargeID)
}

func (l *ledger) len() int {
	l.mu.Lock()
	defer l.mu.Unlock()
	return len(l.rows)
}

func (l *ledger) since(n int) []string {
	l.mu.Lock()
	defer l.mu.Unlock()
	return append([]string(nil), l.rows[n:]...)
}

// ledgerServer speaks just enough HTTP/1.1 to be a fault injector. Raw sockets
// rather than net/http because two of the faults -- an RST, and closing with no
// response at all -- are below what a Handler can express.
type ledgerServer struct {
	listener net.Listener
	led      *ledger
	mu       sync.Mutex
	held     []net.Conn // `hang` connections, closed at shutdown
}

func newLedgerServer(led *ledger) (*ledgerServer, error) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return nil, err
	}
	s := &ledgerServer{listener: ln, led: led}
	go s.acceptLoop()
	return s, nil
}

func (s *ledgerServer) port() int { return s.listener.Addr().(*net.TCPAddr).Port }

func (s *ledgerServer) acceptLoop() {
	for {
		conn, err := s.listener.Accept()
		if err != nil {
			return
		}
		go s.serve(conn)
	}
}

func (s *ledgerServer) serve(conn net.Conn) {
	buf := make([]byte, 4096)
	_ = conn.SetReadDeadline(time.Now().Add(5 * time.Second))
	n, err := conn.Read(buf)
	if err != nil || n == 0 {
		conn.Close()
		return
	}
	line := strings.SplitN(string(buf[:n]), "\r\n", 2)[0]
	fields := strings.Fields(line)
	if len(fields) < 2 {
		conn.Close()
		return
	}
	parts := strings.SplitN(fields[1], "/", 4) // ["", "charge", mode, chargeID]
	if len(parts) < 4 {
		conn.Close()
		return
	}
	mode, chargeID := parts[2], parts[3]

	switch mode {
	case "ok":
		s.led.commit(chargeID)
		reply(conn, chargeID)
		conn.Close()
	case "slow":
		s.led.commit(chargeID)
		time.Sleep(slowResponse)
		reply(conn, chargeID)
		conn.Close()
	case "hang":
		s.led.commit(chargeID)
		s.mu.Lock()
		s.held = append(s.held, conn) // accepted, committed, never answered
		s.mu.Unlock()
	case "reset":
		s.led.commit(chargeID)
		// SetLinger(0) makes Close send RST instead of FIN: what a peer that
		// panics, or a middlebox that gives up, looks like from the client.
		if tcp, ok := conn.(*net.TCPConn); ok {
			_ = tcp.SetLinger(0)
		}
		conn.Close()
	case "crash_after_commit":
		// The case no timeout tuning can fix: durable work, dead reporter.
		s.led.commit(chargeID)
		conn.Close()
	default:
		conn.Close()
	}
}

func reply(conn net.Conn, chargeID string) {
	body := fmt.Sprintf(`{"charge_id":%q}`, chargeID)
	_, _ = io.WriteString(conn, fmt.Sprintf(
		"HTTP/1.1 200 OK\r\nContent-Type: application/json\r\n"+
			"Content-Length: %d\r\nConnection: close\r\n\r\n%s", len(body), body))
}

func (s *ledgerServer) shutdown() {
	s.mu.Lock()
	for _, c := range s.held {
		c.Close()
	}
	s.mu.Unlock()
	s.listener.Close()
}

// --- classification ---------------------------------------------------------

type verdict struct {
	kind  string // SUCCESS, SAFE, AMBIGUOUS
	label string
}

// doRequest issues one request and reports which phase it died in.
//
// The decision is not "is this a timeout" -- it is "were the request bytes
// written". httptrace answers that exactly: WroteRequest fires once the
// transport has finished writing the request. Before it, a retry is provably
// safe. After it, the server may hold a committed charge you will never hear
// about.
func doRequest(client *http.Client, url string) verdict {
	var (
		mu           sync.Mutex
		wroteRequest bool
		connected    bool
	)
	trace := &httptrace.ClientTrace{
		ConnectDone: func(_, _ string, err error) {
			mu.Lock()
			defer mu.Unlock()
			if err == nil {
				connected = true
			}
		},
		WroteRequest: func(info httptrace.WroteRequestInfo) {
			mu.Lock()
			defer mu.Unlock()
			if info.Err == nil {
				wroteRequest = true
			}
		},
	}

	ctx, cancel := context.WithTimeout(context.Background(), clientTimeout)
	defer cancel()
	req, err := http.NewRequestWithContext(
		httptrace.WithClientTrace(ctx, trace), http.MethodGet, url, nil)
	if err != nil {
		return verdict{"AMBIGUOUS", "malformed request"}
	}

	resp, err := client.Do(req)
	if err == nil {
		_, _ = io.Copy(io.Discard, resp.Body)
		resp.Body.Close()
		return verdict{"SUCCESS", fmt.Sprintf("%d", resp.StatusCode)}
	}

	mu.Lock()
	sent, dialed := wroteRequest, connected
	mu.Unlock()

	label := describe(err)
	if !sent {
		// Nothing (or not all) of the request reached the socket. A retry
		// cannot duplicate work.
		if !dialed {
			return verdict{"SAFE", label + " [pre-connect]"}
		}
		return verdict{"SAFE", label + " [pre-write]"}
	}
	return verdict{"AMBIGUOUS", label + " [post-write]"}
}

func describe(err error) string {
	switch {
	case errors.Is(err, syscall.ECONNREFUSED):
		return "ECONNREFUSED"
	case errors.Is(err, syscall.ECONNRESET):
		return "ECONNRESET"
	case errors.Is(err, context.DeadlineExceeded):
		return "DeadlineExceeded"
	case errors.Is(err, io.EOF) || strings.Contains(err.Error(), "EOF"):
		return "EOF"
	}
	var opErr *net.OpError
	if errors.As(err, &opErr) {
		return "net.OpError(" + opErr.Op + ")"
	}
	return "unclassified"
}

// --- phases -----------------------------------------------------------------

type phaseResult struct{ duplicates, unresolved, rows int }

func runPhase(tag, name, note string, port, closedPort int, led *ledger, retryAmbiguous bool) phaseResult {
	before := led.len()
	// A transport that reuses nothing: each attempt gets a fresh connection, so
	// a fault on one cannot be mistaken for a fault on another.
	client := &http.Client{
		Transport: &http.Transport{
			DialContext:       (&net.Dialer{Timeout: clientTimeout}).DialContext,
			DisableKeepAlives: true,
		},
	}

	fmt.Println()
	fmt.Printf("  %s\n  %s\n", name, note)
	fmt.Printf("  %-20s %-44s %9s %12s\n", "fault", "client verdict", "attempts", "ledger rows")

	unresolved := 0
	for _, mode := range modes {
		modeBefore := led.len()
		attempts := 0
		counts := map[string]int{}
		targetPort := port
		if mode == "refused" {
			targetPort = closedPort
		}

		for i := 0; i < requestsPerMode; i++ {
			chargeID := fmt.Sprintf("%s-%s-%d", tag, mode, i)
			url := fmt.Sprintf("http://127.0.0.1:%d/charge/%s/%s", targetPort, mode, chargeID)
			var v verdict
			for a := 0; a < maxAttempts; a++ {
				attempts++
				v = doRequest(client, url)
				if v.kind == "SUCCESS" {
					break
				}
				if v.kind == "SAFE" {
					continue // provably safe: try again
				}
				if retryAmbiguous {
					continue // the bug, made explicit
				}
				break // correct: stop, escalate
			}
			counts[v.kind+"("+v.label+")"]++
			if v.kind == "AMBIGUOUS" {
				unresolved++
			}
		}

		keys := make([]string, 0, len(counts))
		for k := range counts {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			parts = append(parts, fmt.Sprintf("%dx %s", counts[k], k))
		}
		fmt.Printf("  %-20s %-44s %9d %12d\n",
			mode, strings.Join(parts, ", "), attempts, led.len()-modeBefore)
	}

	written := led.since(before)
	seen := map[string]int{}
	for _, id := range written {
		seen[id]++
	}
	duplicates := 0
	for _, n := range seen {
		if n > 1 {
			duplicates += n - 1
		}
	}
	fmt.Printf("  ledger rows written this phase : %d\n", len(written))
	fmt.Printf("  DUPLICATE CHARGES              : %d   <- created by this client's retries\n", duplicates)
	fmt.Printf("  unresolved ambiguous outcomes  : %d   <- caller cannot tell whether these happened\n", unresolved)
	return phaseResult{duplicates, unresolved, len(written)}
}

func findClosedPort() int {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 1
	}
	port := ln.Addr().(*net.TCPAddr).Port
	ln.Close()
	return port
}

func main() {
	led := &ledger{}
	server, err := newLedgerServer(led)
	if err != nil {
		panic(err)
	}
	defer server.shutdown()
	closedPort := findClosedPort()

	fmt.Println(strings.Repeat("=", 78))
	fmt.Println("Layer 4 · Topic 1 — partial failure and the ambiguous result (Go)")
	fmt.Println(strings.Repeat("=", 78))
	fmt.Printf("  ledger        : 127.0.0.1:%d  (in-process, holds server-side truth)\n", server.port())
	fmt.Printf("  closed port   : 127.0.0.1:%d  (for the connect-refused case)\n", closedPort)
	fmt.Printf("  client timeout: %v   slow response: %v   max attempts: %d\n",
		clientTimeout, slowResponse, maxAttempts)
	fmt.Println("  phase detection: net/http/httptrace ConnectDone + WroteRequest")

	naive := runPhase("p1",
		"phase 1 — retry on any error",
		"`if err != nil { retry() }`, which is what most Go clients do",
		server.port(), closedPort, led, true)
	fixed := runPhase("p2",
		"phase 2 — retry only when httptrace proves the request was never written",
		"pre-connect and pre-write failures are retried; everything else escalates",
		server.port(), closedPort, led, false)

	fmt.Println()
	fmt.Println(strings.Repeat("-", 78))
	fmt.Printf("  duplicate charges    phase 1: %-6d phase 2: %d\n", naive.duplicates, fixed.duplicates)
	fmt.Printf("  unresolved ambiguity phase 1: %-6d phase 2: %d\n", naive.unresolved, fixed.unresolved)
	fmt.Println()
	fmt.Println("  Context cancellation is a local event. When the deadline fires, Go tears")
	fmt.Println("  down this side of the connection and returns. The handler on the other")
	fmt.Println("  side has already read the request and finishes its work regardless --")
	fmt.Println("  which is why the ledger row count above keeps rising for faults the")
	fmt.Println("  client recorded as failures. Making that retry safe is Topic 2.")
}

File: 04-distributed/01-partial-failure-and-the-ambiguous-result/cpp/ambiguous_result.cpp
// Layer 4 · Topic 1 — the third outcome, in C++.
//
// WHAT THIS DEMONSTRATES
//   The same five faults and the same ledger as the other five programs. C++ is
//   here because it is the only language in this lab with nothing between the
//   program and the kernel, and this topic is entirely about a boundary the
//   kernel draws:
//
//     connect()  failed  -> not one byte of your request exists anywhere.
//     send()     failed with 0 bytes written -> same.
//     send()     failed after n>0 bytes      -> the peer may have all of it.
//     recv()     failed  -> the request was delivered. You will never know more.
//
//   Every HTTP client in the other five languages is a wrapper that decides how
//   much of that to show you. Here there is no wrapper, so the safe/unsafe
//   decision is not an exception taxonomy to learn -- it is which line returned
//   -1, and what errno says.
//
//   And, unlike Rust: nothing stops you writing the wrong thing. Phase 1's
//   `if (outcome != SUCCESS) continue;` compiles without a warning at -Wall
//   -Wextra. That is the comparison worth making.
//
// WHAT TO LOOK FOR
//   Phase 1's duplicate charges against phase 2's, and the unresolved ambiguity
//   that survives both.
//
// Build & run:
//   g++ -O2 -std=c++17 -pthread -Wall -Wextra -o /tmp/l4t1_cpp cpp/ambiguous_result.cpp && /tmp/l4t1_cpp

// NOTE: htonl/htons/ntohs are function-like MACROS on Darwin (sys/_endian.h),
// so they must not be written as ::htonl(...). On glibc they are functions and
// the :: form compiles, which is exactly how a Linux-only file fails here.
#include <arpa/inet.h>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstring>
#include <errno.h>
#include <fcntl.h>
#include <map>
#include <mutex>
#include <netinet/in.h>
#include <poll.h>
#include <string>
#include <sys/socket.h>
#include <thread>
#include <unistd.h>
#include <vector>

static const int CLIENT_TIMEOUT_MS = 300;
static const int SLOW_RESPONSE_MS = 1000;
static const int REQUESTS_PER_MODE = 4;
static const int MAX_ATTEMPTS = 3;

static const char* MODES[] = {"ok", "slow", "hang", "reset",
                              "crash_after_commit", "refused"};
static const int MODE_COUNT = 6;

// --- server-side truth ------------------------------------------------------

static std::mutex g_ledger_mutex;
static std::vector<std::string> g_ledger;

static void commit(const std::string& charge_id) {
    std::lock_guard<std::mutex> lock(g_ledger_mutex);
    g_ledger.push_back(charge_id);
}

static size_t ledger_size() {
    std::lock_guard<std::mutex> lock(g_ledger_mutex);
    return g_ledger.size();
}

static std::mutex g_held_mutex;
static std::vector<int> g_held_fds;  // `hang` connections, closed at shutdown

static void send_reply(int fd, const std::string& charge_id) {
    std::string body = "{\"charge_id\":\"" + charge_id + "\"}";
    char head[256];
    int n = std::snprintf(head, sizeof(head),
                          "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\n"
                          "Content-Length: %zu\r\nConnection: close\r\n\r\n",
                          body.size());
    (void)::send(fd, head, (size_t)n, 0);
    (void)::send(fd, body.data(), body.size(), 0);
}

static void serve_connection(int fd) {
    char buf[4096];
    ssize_t n = ::recv(fd, buf, sizeof(buf) - 1, 0);
    if (n <= 0) {
        ::close(fd);
        return;
    }
    buf[n] = '\0';

    // "GET /charge/<mode>/<charge_id> HTTP/1.1"
    std::string request(buf);
    size_t sp1 = request.find(' ');
    size_t sp2 = request.find(' ', sp1 + 1);
    if (sp1 == std::string::npos || sp2 == std::string::npos) {
        ::close(fd);
        return;
    }
    std::string path = request.substr(sp1 + 1, sp2 - sp1 - 1);
    size_t s2 = path.find('/', 1);
    size_t s3 = path.find('/', s2 + 1);
    if (s2 == std::string::npos || s3 == std::string::npos) {
        ::close(fd);
        return;
    }
    std::string mode = path.substr(s2 + 1, s3 - s2 - 1);
    std::string charge_id = path.substr(s3 + 1);

    if (mode == "ok") {
        commit(charge_id);
        send_reply(fd, charge_id);
        ::close(fd);
    } else if (mode == "slow") {
        commit(charge_id);
        std::this_thread::sleep_for(std::chrono::milliseconds(SLOW_RESPONSE_MS));
        send_reply(fd, charge_id);
        ::close(fd);
    } else if (mode == "hang") {
        // Accepted, committed, never answered. The fd is parked rather than
        // closed, because closing it is what the client is waiting for.
        commit(charge_id);
        std::lock_guard<std::mutex> lock(g_held_mutex);
        g_held_fds.push_back(fd);
    } else if (mode == "reset") {
        commit(charge_id);
        // SO_LINGER with a zero timeout turns close() into an RST rather than a
        // FIN. This is what a peer that aborts, or a middlebox that gives up,
        // looks like from the caller's side.
        struct linger lin;
        lin.l_onoff = 1;
        lin.l_linger = 0;
        ::setsockopt(fd, SOL_SOCKET, SO_LINGER, &lin, sizeof(lin));
        ::close(fd);
    } else if (mode == "crash_after_commit") {
        // The case no timeout tuning can fix: durable work, dead reporter.
        commit(charge_id);
        ::close(fd);
    } else {
        ::close(fd);
    }
}

static int start_ledger_server() {
    int listen_fd = ::socket(AF_INET, SOCK_STREAM, 0);
    int one = 1;
    ::setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
    struct sockaddr_in addr;
    std::memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = 0;  // let the kernel choose
    ::bind(listen_fd, (struct sockaddr*)&addr, sizeof(addr));
    ::listen(listen_fd, 128);

    std::thread([listen_fd] {
        for (;;) {
            int fd = ::accept(listen_fd, nullptr, nullptr);
            if (fd < 0) return;
            std::thread(serve_connection, fd).detach();
        }
    }).detach();
    return listen_fd;
}

static int port_of(int fd) {
    struct sockaddr_in addr;
    socklen_t len = sizeof(addr);
    ::getsockname(fd, (struct sockaddr*)&addr, &len);
    return ntohs(addr.sin_port);
}

// --- the client, one phase at a time ----------------------------------------

enum Kind { SUCCESS, SAFE, AMBIGUOUS };

struct Outcome {
    Kind kind;
    std::string label;
};

static Outcome ok_outcome(int code) {
    return {SUCCESS, "SUCCESS(" + std::to_string(code) + ")"};
}
static Outcome safe(const char* phase, int err) {
    return {SAFE, std::string("SAFE(") + std::strerror(err) + " [" + phase + "])"};
}
static Outcome unknown(const char* phase, const std::string& detail) {
    return {AMBIGUOUS, std::string("AMBIGUOUS(") + detail + " [" + phase + "])"};
}

// Non-blocking connect + poll, because connect(2) does not honour SO_SNDTIMEO.
// Returns 0 on success, otherwise sets errno.
static int connect_with_timeout(int fd, const struct sockaddr_in& addr, int timeout_ms) {
    int flags = ::fcntl(fd, F_GETFL, 0);
    ::fcntl(fd, F_SETFL, flags | O_NONBLOCK);

    int rc = ::connect(fd, (const struct sockaddr*)&addr, sizeof(addr));
    if (rc == 0) {
        ::fcntl(fd, F_SETFL, flags);
        return 0;
    }
    if (errno != EINPROGRESS) return -1;

    struct pollfd pfd;
    pfd.fd = fd;
    pfd.events = POLLOUT;
    rc = ::poll(&pfd, 1, timeout_ms);
    if (rc == 0) {
        errno = ETIMEDOUT;
        return -1;
    }
    if (rc < 0) return -1;

    int sock_err = 0;
    socklen_t len = sizeof(sock_err);
    ::getsockopt(fd, SOL_SOCKET, SO_ERROR, &sock_err, &len);
    ::fcntl(fd, F_SETFL, flags);
    if (sock_err != 0) {
        errno = sock_err;
        return -1;
    }
    return 0;
}

static Outcome attempt(int port, const std::string& path) {
    int fd = ::socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) return unknown("socket", std::strerror(errno));

    struct sockaddr_in addr;
    std::memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = htons((uint16_t)port);

    // PHASE 1 - connect. If this fails, nothing was sent. Provably safe.
    if (connect_with_timeout(fd, addr, CLIENT_TIMEOUT_MS) != 0) {
        int e = errno;
        ::close(fd);
        return safe("connect", e);
    }

    struct timeval tv;
    tv.tv_sec = CLIENT_TIMEOUT_MS / 1000;
    tv.tv_usec = (CLIENT_TIMEOUT_MS % 1000) * 1000;
    ::setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
    ::setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));

    // PHASE 2 - send. Zero bytes written is provably safe; a partial write is
    // not, because the peer may already hold a complete request.
    std::string request = "GET " + path +
                          " HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\n\r\n";
    size_t written = 0;
    while (written < request.size()) {
        ssize_t n = ::send(fd, request.data() + written, request.size() - written, 0);
        if (n > 0) {
            written += (size_t)n;
            continue;
        }
        int e = errno;
        if (n < 0 && e == EINTR) continue;
        ::close(fd);
        if (written == 0) return safe("send", e);
        return unknown("partial-send", std::strerror(e));
    }

    // PHASE 3 - receive. Anything that goes wrong from here is unknowable: the
    // request was delivered, and the outcome lives on the far side of whatever
    // just broke.
    std::string response;
    char buf[4096];
    for (;;) {
        ssize_t n = ::recv(fd, buf, sizeof(buf), 0);
        if (n > 0) {
            response.append(buf, (size_t)n);
            continue;
        }
        if (n == 0) break;  // clean EOF
        if (errno == EINTR) continue;
        int e = errno;
        ::close(fd);
        if (e == EAGAIN || e == EWOULDBLOCK) return unknown("recv", "timeout");
        return unknown("recv", std::strerror(e));
    }
    ::close(fd);

    if (response.empty()) return unknown("recv", "closed with no response");
    int code = 0;
    if (std::sscanf(response.c_str(), "HTTP/1.%*d %d", &code) != 1)
        return unknown("recv", "unparseable response");
    return ok_outcome(code);
}

// --- phases -----------------------------------------------------------------

struct PhaseResult {
    int duplicates;
    int unresolved;
};

static PhaseResult run_phase(const char* tag, const char* name, const char* note,
                             int port, int closed_port, bool retry_ambiguous) {
    size_t before = ledger_size();
    int unresolved = 0;

    std::printf("\n  %s\n  %s\n", name, note);
    std::printf("  %-20s %-44s %9s %12s\n", "fault", "client verdict", "attempts",
                "ledger rows");

    for (int m = 0; m < MODE_COUNT; m++) {
        const std::string mode = MODES[m];
        size_t mode_before = ledger_size();
        int attempts = 0;
        std::map<std::string, int> counts;
        int target = (mode == "refused") ? closed_port : port;

        for (int i = 0; i < REQUESTS_PER_MODE; i++) {
            std::string charge_id = std::string(tag) + "-" + mode + "-" + std::to_string(i);
            std::string path = "/charge/" + mode + "/" + charge_id;
            Outcome outcome{AMBIGUOUS, "AMBIGUOUS(not attempted [none])"};
            for (int a = 0; a < MAX_ATTEMPTS; a++) {
                attempts++;
                outcome = attempt(target, path);
                if (outcome.kind == SUCCESS) break;
                if (outcome.kind == SAFE) continue;   // provably safe: try again
                if (retry_ambiguous) continue;        // the bug, made explicit
                break;                                // correct: stop, escalate
            }
            if (outcome.kind == AMBIGUOUS) unresolved++;
            counts[outcome.label]++;
        }

        std::string summary;
        for (const auto& kv : counts) {
            if (!summary.empty()) summary += ", ";
            summary += std::to_string(kv.second) + "x " + kv.first;
        }
        std::printf("  %-20s %-44s %9d %12zu\n", mode.c_str(), summary.c_str(),
                    attempts, ledger_size() - mode_before);
    }

    std::map<std::string, int> seen;
    {
        std::lock_guard<std::mutex> lock(g_ledger_mutex);
        for (size_t i = before; i < g_ledger.size(); i++) seen[g_ledger[i]]++;
    }
    int duplicates = 0;
    int rows = 0;
    for (const auto& kv : seen) {
        rows += kv.second;
        if (kv.second > 1) duplicates += kv.second - 1;
    }
    std::printf("  ledger rows written this phase : %d\n", rows);
    std::printf("  DUPLICATE CHARGES              : %d   <- created by this client's retries\n",
                duplicates);
    std::printf("  unresolved ambiguous outcomes  : %d   <- caller cannot tell whether these happened\n",
                unresolved);
    return {duplicates, unresolved};
}

static int find_closed_port() {
    int fd = ::socket(AF_INET, SOCK_STREAM, 0);
    struct sockaddr_in addr;
    std::memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = 0;
    ::bind(fd, (struct sockaddr*)&addr, sizeof(addr));
    int port = port_of(fd);
    ::close(fd);
    return port;
}

int main() {
    // Writing to a socket the peer has RST'd raises SIGPIPE, whose default
    // action is to kill the process. Every network program in C has to do this
    // and it is a genuine production outage waiting for the day a dependency
    // starts resetting connections.
    std::signal(SIGPIPE, SIG_IGN);

    int listen_fd = start_ledger_server();
    int port = port_of(listen_fd);
    int closed_port = find_closed_port();

    std::string bar(78, '=');
    std::printf("%s\nLayer 4 · Topic 1 — partial failure and the ambiguous result (C++)\n%s\n",
                bar.c_str(), bar.c_str());
    std::printf("  ledger        : 127.0.0.1:%d  (in-process, holds server-side truth)\n", port);
    std::printf("  closed port   : 127.0.0.1:%d  (for the connect-refused case)\n", closed_port);
    std::printf("  client timeout: %dms   slow response: %dms   max attempts: %d\n",
                CLIENT_TIMEOUT_MS, SLOW_RESPONSE_MS, MAX_ATTEMPTS);
    std::printf("  phase detection: which syscall returned -1\n");

    PhaseResult naive = run_phase("p1", "phase 1 — retry on any error",
                                  "`if (rc != 0) continue;` -- compiles clean at -Wall -Wextra",
                                  port, closed_port, true);
    PhaseResult fixed = run_phase("p2", "phase 2 — retry only connect/zero-byte-send failures",
                                  "the only two cases where the kernel proves nothing was sent",
                                  port, closed_port, false);

    std::string dash(78, '-');
    std::printf("\n%s\n", dash.c_str());
    std::printf("  duplicate charges    phase 1: %-6d phase 2: %d\n", naive.duplicates,
                fixed.duplicates);
    std::printf("  unresolved ambiguity phase 1: %-6d phase 2: %d\n", naive.unresolved,
                fixed.unresolved);
    std::printf(
        "\n  Nothing in this file prevented phase 1. Compare the Rust program, where\n"
        "  the same mistake is a compile error. Same hazard, same fix, and the only\n"
        "  difference is whether the language will let you ship it.\n");

    {
        std::lock_guard<std::mutex> lock(g_held_mutex);
        for (int fd : g_held_fds) ::close(fd);
    }
    ::close(listen_fd);
    return 0;
}
